miso-web/src/main/webapp/scripts/library_hot.js

```diff
@@ -2,11 +2,812 @@ Library.hot = {
+    xhr.onreadystatechange = function () {
+      if (xhr.readyState === XMLHttpRequest.DONE) {
+        callback(); // Indicate request has completed.
+        xhr.status === 200 ? Library.hot.successSave(xhr, rowIndex, numberToSave) : Library.hot.failSave(xhr, rowIndex, numberToSave);
+      }
+    };
+    xhr.open('PUT', '/miso/rest/library/' + libraryId);
+    xhr.setRequestHeader('Content-Type', 'application/json');
+    xhr.send(data);
+  },
+  
+  failSave: function (xhr, rowIndex, numberToSave) {
+    console.log(xhr);
+    var responseText = JSON.parse(xhr.responseText);
+    var allColumnData = Library.hot.getValues('data', Library.hot.colConf);
+    var column, columnIndex;
+    if (responseText.data && responseText.data.constraintName) {
+      // if a column's constraint was violated, extract it here
+      column = responseText.data.constraintName;
+      columnIndex = allColumnData.indexOf(column);
+    }
+    console.log(rowIndex, columnIndex);
+    if (rowIndex !== undefined && columnIndex !== -1 && columnIndex !== undefined) {
+      Library.hot.hotTable.setCellMeta(rowIndex, columnIndex, 'valid', false);
+    }
+    // process error message if it was a SQL violation, and add any errors to the messages array
+    var reUserMessage = /could not execute .*?: (.*)/;
+    Library.hot.messages.failed.push("Row "+ (rowIndex + 1) +": "+ responseText.detail.replace(reUserMessage, "$1")); 
+    
+    // display any errors if this is the final sample to be saved
+    if (Library.hot.messages.success.length + Library.hot.messages.failed.length == numberToSave) {
+      Library.hot.addAnyErrors();
+    }
+  },
+  
+  successSave: function (xhr, rowIndex, numberToSave) {
+    // push row index for new saves (previously-saved items have their aliases added)
+    Library.hot.messages.success.push(rowIndex); 
+    
+    // add a 'saved' attribute to the data source 
+    Library.hot.startData[rowIndex].saved = true;
+    
+    // add sample url and id to the data source if the sample is newly created
+    if (!Library.hot.startData[rowIndex].id) {
+      Library.hot.startData[rowIndex].url = xhr.getResponseHeader('Location');
+      Library.hot.startData[rowIndex].id = Library.hot.startData[rowIndex].url.split('/').pop();
+    }
+    
+    // display any errors if this is the final sample to be saved
+    if (Library.hot.messages.success.length + Library.hot.messages.failed.length == numberToSave) {
+      Library.hot.addAnyErrors();
+    }
+  },
+  
+  createData: function () {
+    var continueValidation = Library.hot.cleanRowsAndToggleSaveButton();
+    if (continueValidation === false) return false;
+    
+    Library.hot.hotTable.validateCells(function (isValid) {
+      if (isValid) {
+        // send it through the parser to get a sampleData array that isn't merely a reference to Library.hot.hotTable.getSourceData()
+        var libsData = JSON.parse(JSON.parse(JSON.stringify(Library.hot.hotTable.getSourceData())));
```

Comment:
I can't believe this is the best way, but StackOverflow says it, so it must be true. Should there be two `parse`s?